### PR TITLE
Smarter update strategy for Onix report [GAWB-2963]

### PR DESCRIPTION
**Pull request by aednichols**
~Needs unit tests, probably fix up some of the clunky naming. Added~ There are still two requirements questions to the JIRA ticket, so may need updates for those.

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've followed [the instructions](https://github.com/broadinstitute/firecloud-orchestration/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/broadinstitute/firecloud-orchestration/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this **unless it's for a hotfix**. In that case, don't delete it!
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala [hunk 1]
```diff
@@ -334,16 +334,53 @@ object HttpGoogleServicesDAO extends GoogleServicesDAO with FireCloudRequestBuil
...
+  private def updatePreservingOrder(newContent: ValueRange, existingContent: ValueRange): List[java.util.List[AnyRef]] = {
+    val header: java.util.List[AnyRef] = existingContent.getValues.get(0)
+    val existingRecords: List[java.util.List[AnyRef]] = existingContent.getValues.drop(1).toList
+
+    val newRecords: List[java.util.List[AnyRef]] = newContent.getValues.drop(1).toList
+
+    // Go through existing records and update them in place
+    val existingRecordsUpdated = existingRecords.map { existingRecord =>
+      val matchingNewRecord = newRecords.find { newRecordCandidate =>
+        newRecordCandidate.get(0) == existingRecord.get(0)
```
Comment (aednichols) [inline src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala]: I don't really like using `.get(0)` everywhere to access the project ID.

I was thinking of refactoring `SpreadsheetRow` to include it, and adopting that class here, but I'm not sure it's worth it.

File: src/test/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAOSpec.scala [hunk 2]
```diff
@@ -63,4 +65,32 @@ class HttpGoogleServicesDAOSpec extends FlatSpec with Matchers {
     objectMetadata.name should equal("test-composite-object")
     objectMetadata.md5Hash should equal(None)
   }
+
+  {
+    // https://stackoverflow.com/a/24375762/818054
+    val updatePreservingOrder = PrivateMethod('updatePreservingOrder)
```
Comment (breilly2) [inline src/test/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAOSpec.scala]: Does `updatePreservingOrder` have to be private? Or can you extract the thing that you want to test into a separate object that this can logically include in its public API?

Comment (aednichols) [reply src/test/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAOSpec.scala]: Good question. The public API is `updateSpreadsheet`, above, which interacts directly with Google - so we can't conveniently test that.

I'd entertain making `updatePreservingOrder` `public` to make tests simpler, but it doesn't naturally belong to the public API of `HttpGoogleServicesDAO`.

This is the first time I've encountered this specific problem in Scala, and decided to explore what's possible in `ScalaTest`.

File: src/test/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAOSpec.scala [hunk 3]
```diff
@@ -63,4 +65,86 @@ class HttpGoogleServicesDAOSpec extends FlatSpec with Matchers {
...
+      check(
+        newContent = List(headers, row2, row3, row1),
+        existingContent = List(headers, row1, row2, row3),
+        expectedOutput = List(headers, row1, row2, row3)
+      )
+    }
+
+    def check(newContent: List[java.util.List[AnyRef]], existingContent: List[java.util.List[AnyRef]], expectedOutput: List[java.util.List[AnyRef]]): Unit = {
+      // https://stackoverflow.com/a/24375762/818054
+      val updatePreservingOrder = PrivateMethod('updatePreservingOrder)
```
Comment (aednichols) [inline src/test/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAOSpec.scala:193]: If people don't like this (and I can see why they wouldn't), happy to just make `updatePreservingOrder` `public`

File: src/test/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAOSpec.scala [hunk 4]
```diff
@@ -63,4 +65,102 @@ class HttpGoogleServicesDAOSpec extends FlatSpec with Matchers {
...
+        expectedOutput = List(headers, row1, row2, row3)
+      )
+    }
+
+    it should "output the union of the rows (disjoint sets)" in {
+      check(
+        newContent = List(headers, row3),
+        existingContent = List(headers, row1, row2),
+        expectedOutput = List(headers, row1, row2, row3)
+      )
```
Comment (davidangb) [inline src/test/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAOSpec.scala:148]: I think these last few are order-dependent, aren't they? Might be worth mucking with the order a bit ... 

```
 check(
        newContent = List(headers, row1),
        existingContent = List(headers, row2, row3),
        expectedOutput = List(headers, row2, row3, row1)
      )
```

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala [hunk 5]
```diff
@@ -334,16 +334,53 @@ object HttpGoogleServicesDAO extends GoogleServicesDAO with FireCloudRequestBuil
...
+    // Create new records for newly-added projects
+    val recordsToAppend = newRecords.filter { newRecordCandidate =>
+      !existingRecords.exists { existingRecordCandidate =>
+        existingRecordCandidate.get(0) == newRecordCandidate.get(0)
+      }
+    }
+
+    List(header) ++ existingRecordsUpdated ++ recordsToAppend
+  }
+
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala:388]: I would have transformed both `newRecords` and `existingRecords` into maps keyed by the project id: `Map[String, List[AnyRef]]`. Then, inside your loops, instead of doing a `find` or `exists`, you can do a more-efficient map lookup.

At the scale we're talking about - getting up to a thousand-ish rows - this may be premature optimization. So don't consider it a requirement for this PR; I'm mentioning it for the sake of talking about code :)

**Review by davidangb — APPROVED** [review]
:+1:, optional comments inline

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala (hunk 1 above)
Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala]: How about something more scala-y, but the same thing ... `newRecordCandidate.head == existingRecord.head` ?

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala [hunk 6]
```diff
@@ -334,16 +334,53 @@ object HttpGoogleServicesDAO extends GoogleServicesDAO with FireCloudRequestBuil
...
+    // Smart merge existing with new
+    val rows = updatePreservingOrder(newContent, existingContent)
+
+    // Send update
+    val response = service.spreadsheets().values().update(spreadsheetId, newContent.getRange, newContent.setValues(rows)).setValueInputOption("RAW").execute()
     response.toString.parseJson.asJsObject
   }
 
+  private def updatePreservingOrder(newContent: ValueRange, existingContent: ValueRange): List[java.util.List[AnyRef]] = {
+    val header: java.util.List[AnyRef] = existingContent.getValues.get(0)
```
Comment (rushtong) [inline src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala]: Consider `existingContent.getValues.head` if you keep this approach

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala (hunk 5 above)
Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala:388]: +1 to this idea.

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala [hunk 7]
```diff
@@ -334,16 +334,53 @@ object HttpGoogleServicesDAO extends GoogleServicesDAO with FireCloudRequestBuil
...
+    val newRecords: List[java.util.List[AnyRef]] = newContent.getValues.drop(1).toList
+
+    // Go through existing records and update them in place
+    val existingRecordsUpdated = existingRecords.map { existingRecord =>
+      val matchingNewRecord = newRecords.find { newRecordCandidate =>
+        newRecordCandidate.get(0) == existingRecord.get(0)
+      }
+
+      matchingNewRecord match {
+        case Some(newRecord) => newRecord // Overwrite the entire row in place, adequate unless we want to preserve user-added columns
```
Comment (rushtong) [inline src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala]: I do think we should consider preserving user-added columns, not necessarily in this PR. One of the reasons for keeping a data row in the same order would be this exact case - the user has added a new column with a comment that applies to the current row. If we are not trying to preserve user-comments in additional columns, then there's no real point in preserving the order of the data rows. 

That being said, I think this might still work (will test functionally). A new `ValueRange`, when pushed to the spreadsheet, *only* overwrites values *inside* the range and leaves values (and formats, etc.) outside the range intact.

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala [hunk 8]
```diff
@@ -334,16 +334,53 @@ object HttpGoogleServicesDAO extends GoogleServicesDAO with FireCloudRequestBuil
...
+        case _ => existingRecord // Don't delete existing records that aren't in the new export
```
Comment (rushtong) [inline src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala:375]: I am conflicted about this. I agree that we should not try to delete rows. However, I think we do need to indicate to the user that FireCloud no longer knows about this particular row. How we do that is an open question ... perhaps a separate story/PR.

File: src/test/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAOSpec.scala (hunk 3 above)
Comment (rushtong) [reply src/test/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAOSpec.scala:193]: Cool! I didn't know this was possible.

**Review by rushtong — APPROVED** [review]
👍 For code. Will do some functional testing as well.

File: src/test/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAOSpec.scala (hunk 2 above)
Comment (rushtong) [reply src/test/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAOSpec.scala]: Another thought is to pull the method into a supporting trait that would allow for easier testing.

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala (hunk 6 above)
Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala]: In functional testing, `existingContent.getValues` can be null for brand new reports. This method needs to handle both cases, new + empty, existing + filled.

Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala]: Minor issue. In functional testing, headers are maintained and not changed. So, if I have an older spreadsheet that is missing a newer column, then update, the new column data is written without a header.

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala (hunk 7 above)
Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala]: Update. I believe this is functioning as it should: 
* New records do not overwrite data in user-added columns and instead are appended as new rows. 
* Existing records are overwritten in place without changing user-added columns outside of the updated data. The caveat to this is that if new columns are added to the output beyond what the existing spreadsheet has, they *will* overwrite the user-added columns that exist in the same position.

**Review by rushtong — CHANGES_REQUESTED** [review]
Need to handle the empty data case in `updatePreservingOrder`

**Review by andy7i — COMMENTED** [review]
Only reviewed test cases:
Overall, this is awesome. I would recommend the following:
1. test for new columns as greg found.
2. test for a data cell being deleted (emptied).
3. test for cases where rows are updated (similar to your tests for sets) with modified rows (row1a, row2a, row3a).

@andy7i Adam edit:
1. `"use the header from the new data in the update"`
2. modified `row2` and `row3` to have a `null` and a `""`, other tests use the data in ways that cover this case
3. `"update data in a row"` and `"locate a row in existing that moved based on its first column, and replace its remaining columns with new"`

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala (hunk 6 above)
Comment (aednichols) [reply src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala]: 1. Done
2. Fixed + added test
3. Fixed + added test

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala (hunk 5 above)
Comment (aednichols) [reply src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala:388]: This is true, I was lazy and optimized for simplicity on the first pass.

The above would need things like `record.head.asInstanceOf[String]` to wrangle row data from `java.util.List[java.util.List[Object]]` into the right types, and I kinda felt like leaving them alone

**Review by davidangb — APPROVED** [review]
:+1: did not test locally, looks good!

File: src/test/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAOSpec.scala [hunk 9]
```diff
@@ -63,4 +65,140 @@ class HttpGoogleServicesDAOSpec extends FlatSpec with Matchers {
...
+
+    it should "not explode if there is no data (only headers)" in {
+      check(
+        newContent = List(headers),
+        existingContent = List(headers),
+        expectedOutput = List(headers)
+      )
+    }
+
+    // This happens when calling updateBillingReport the first time from within createBillingReport
```
Comment (andy7i) [inline src/test/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAOSpec.scala:94]: This is great and makes me think of another edge case: 
What would happen if `updateBillingReport` or `createBillingReport` were called when there is no data yet. If this is a valid scenario, it should handle when newContent==null and existingContent==null (or is an empty list)

Comment (davidangb) [reply src/test/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAOSpec.scala:94]: agree this is a valid edge case, but wonder if it's worth expending effort on. Remember that very few people - only campaign managers, and then a subset of those - will call this, so we can be pretty tolerant of unexpected errors.

Comment (aednichols) [reply src/test/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAOSpec.scala:94]: It's possible, but I'm comfortable leaving the behavior undefined, a.k.a. "don't do that"

**Review by andy7i — APPROVED** [review]
👍 to test plan

**Review by rushtong — APPROVED** [review]
👍 Functionally beautiful!
